cpu/atxmega/periph/uart.c

```diff
@@ -0,0 +1,430 @@
+
+    /* Precision goal was not met calculate baudrate with uart frequency doubled */
+    precision = _xmega_bsel_bscale(&fcpu, &baud, 1, bsel, bscale );
+    *clk2x = 1;
+
+    return (precision - 10000);
+}
+
+static inline void _configure_pins(uart_t uart)
+{
+    /* configure RX pin */
+    if (gpio_is_valid(uart_config[uart].rx_pin)) {
+        gpio_init(uart_config[uart].rx_pin, GPIO_IN);
+    }
+
+    /* configure TX pin */
+    if (gpio_is_valid(uart_config[uart].tx_pin)) {
+        gpio_set(uart_config[uart].tx_pin);
+        gpio_init(uart_config[uart].tx_pin, GPIO_OUT);
+    }
+
+#ifdef MODULE_PERIPH_UART_HW_FC
+    /* TODO */
+#endif
+}
+
+int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
+{
+    int8_t bscale;
+    uint8_t clk2x;
+    uint16_t bsel;
+
+    /* make sure the given device is valid */
+    if (uart >= UART_NUMOF) {
+        return UART_NODEV;
+    }
+
+    uint16_t count = UINT16_MAX;
+    while (avr8_is_uart_tx_pending() && count--) {}
+
+    /* register interrupt context */
+    isr_ctx[uart].rx_cb = rx_cb;
+    isr_ctx[uart].arg = arg;
+
+    /* disable and reset UART */
+    dev(uart)->CTRLA = 0;
+    dev(uart)->CTRLB = 0;
+    dev(uart)->CTRLC = 0;
+
+    _configure_pins(uart);
+
+    /* configure UART to 8N1 mode */
+    dev(uart)->CTRLC = USART_CMODE_ASYNCHRONOUS_gc
+                     | USART_PMODE_DISABLED_gc
+                     | USART_CHSIZE_8BIT_gc;
+
+    /* set clock divider */
+    _xmega_calculate_bsel_bscale(CLOCK_CORECLOCK, baudrate, &clk2x,
+                                 &bsel, &bscale);
+
+    dev(uart)->BAUDCTRLA = (uint8_t)(bsel & 0x00ff);
+    dev(uart)->BAUDCTRLB = (bscale << USART_BSCALE_gp)
+                         | ((uint8_t)((bsel & 0x0fff) >> 8));
+    if (clk2x == 1) {
+        dev(uart)->CTRLB |= USART_CLK2X_bm;
+    }
+
+    /* enable RX and TX Interrupts and set level*/
+    if (rx_cb) {
+        dev(uart)->CTRLA = (uart_config[uart].rx_int_lvl << USART_RXCINTLVL_gp)
+                         | (uart_config[uart].tx_int_lvl << USART_TXCINTLVL_gp)
+                         | (uart_config[uart].dre_int_lvl << USART_DREINTLVL_gp);
+        dev(uart)->CTRLB = USART_RXEN_bm | USART_TXEN_bm;
+    }
+    else {
+        /* only transmit */
+        dev(uart)->CTRLB = USART_TXEN_bm;
+    }
+
+    DEBUG("Set clk2x %" PRIu8 " bsel %" PRIu16 "bscale %" PRIi8 "\n",
+          clk2x, bsel, bscale);
+
+    return UART_OK;
+}
+
+void uart_write(uart_t uart, const uint8_t *data, size_t len)
+{
+    pm_block(3);
+    for (size_t i = 0; i < len; i++) {
+        while (!(dev(uart)->STATUS & USART_DREIF_bm)) {}
+
+        /* start of TX won't finish until no data in DATAn and transmit shift
+           register is empty */
+        uint8_t irq_state = irq_disable();
+        avr8_state |= AVR8_STATE_FLAG_UART_TX(uart);
+        irq_restore(irq_state);
+
+        dev(uart)->DATA = data[i];
+    }
+    while (!(dev(uart)->STATUS & USART_DREIF_bm)) {};
+    pm_unblock(3);
```

Comment:
Note that low power mode is only entered if all threads are currently blocked. In the above code, there is no blocking function called between blocking and unblocking low power mode.

I think the issue here is that the periph is still busy shifting out the bits after the CPU has initialized the transmission of a char. If all threads block after let's say 5 bits have been shifted out, the remaining 3 bits get currupted.